src/script/lua_api/l_object.cpp

```diff
@@ -162,78 +164,67 @@ int ObjectRef::l_punch(lua_State *L)
 	NO_MAP_LOCK_REQUIRED;
 	ObjectRef *ref = checkobject(L, 1);
 	ObjectRef *puncher_ref = checkobject(L, 2);
-	ServerActiveObject *co = getobject(ref);
+	ServerActiveObject *sao = getobject(ref);
 	ServerActiveObject *puncher = getobject(puncher_ref);
-	if (!co || !puncher)
+	if (sao == nullptr || puncher == nullptr)
 		return 0;
-	v3f dir;
-	if (lua_type(L, 5) != LUA_TTABLE)
-		dir = co->getBasePosition() - puncher->getBasePosition();
-	else
-		dir = read_v3f(L, 5);
-	float time_from_last_punch = 1000000;
-	if (lua_isnumber(L, 3))
-		time_from_last_punch = lua_tonumber(L, 3);
+
+	float time_from_last_punch = lua_isnil(L, 3) ?
+		1000000.0f : readParam<float>(L,3);
 	ToolCapabilities toolcap = read_tool_capabilities(L, 4);
-	dir.normalize();
+	v3f dir = lua_isnil(L, 5) ?
+		sao->getBasePosition() - puncher->getBasePosition() : check_v3f(L, 5);
```

Comment:
https://github.com/minetest-mods/throwing_arrows/issues/6

`lua_isnoneornil` is perhaps a better choice here to ensure mod compatibility.